"use strict";(self.webpackChunk=self.webpackChunk||[]).push([[65860],{603905:(e,r,o)=>{o.d(r,{Zo:()=>c,kt:()=>u});var n=o(667294);function a(e,r,o){return r in e?Object.defineProperty(e,r,{value:o,enumerable:!0,configurable:!0,writable:!0}):e[r]=o,e}function t(e,r){var o=Object.keys(e);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(e);r&&(n=n.filter((function(r){return Object.getOwnPropertyDescriptor(e,r).enumerable}))),o.push.apply(o,n)}return o}function l(e){for(var r=1;r<arguments.length;r++){var o=null!=arguments[r]?arguments[r]:{};r%2?t(Object(o),!0).forEach((function(r){a(e,r,o[r])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(o)):t(Object(o)).forEach((function(r){Object.defineProperty(e,r,Object.getOwnPropertyDescriptor(o,r))}))}return e}function i(e,r){if(null==e)return{};var o,n,a=function(e,r){if(null==e)return{};var o,n,a={},t=Object.keys(e);for(n=0;n<t.length;n++)o=t[n],r.indexOf(o)>=0||(a[o]=e[o]);return a}(e,r);if(Object.getOwnPropertySymbols){var t=Object.getOwnPropertySymbols(e);for(n=0;n<t.length;n++)o=t[n],r.indexOf(o)>=0||Object.prototype.propertyIsEnumerable.call(e,o)&&(a[o]=e[o])}return a}var d=n.createContext({}),s=function(e){var r=n.useContext(d),o=r;return e&&(o="function"==typeof e?e(r):l(l({},r),e)),o},c=function(e){var r=s(e.components);return n.createElement(d.Provider,{value:r},e.children)},p={inlineCode:"code",wrapper:function(e){var r=e.children;return n.createElement(n.Fragment,{},r)}},m=n.forwardRef((function(e,r){var o=e.components,a=e.mdxType,t=e.originalType,d=e.parentName,c=i(e,["components","mdxType","originalType","parentName"]),m=s(o),u=a,k=m["".concat(d,".").concat(u)]||m[u]||p[u]||t;return o?n.createElement(k,l(l({ref:r},c),{},{components:o})):n.createElement(k,l({ref:r},c))}));function u(e,r){var o=arguments,a=r&&r.mdxType;if("string"==typeof e||a){var t=o.length,l=new Array(t);l[0]=m;var i={};for(var d in r)hasOwnProperty.call(r,d)&&(i[d]=r[d]);i.originalType=e,i.mdxType="string"==typeof e?e:a,l[1]=i;for(var s=2;s<t;s++)l[s]=o[s];return n.createElement.apply(null,l)}return n.createElement.apply(null,o)}m.displayName="MDXCreateElement"},109842:(e,r,o)=>{o.r(r),o.d(r,{assets:()=>s,contentTitle:()=>i,default:()=>m,frontMatter:()=>l,metadata:()=>d,toc:()=>c});o(667294);var n=o(603905);function a(){return a=Object.assign||function(e){for(var r=1;r<arguments.length;r++){var o=arguments[r];for(var n in o)Object.prototype.hasOwnProperty.call(o,n)&&(e[n]=o[n])}return e},a.apply(this,arguments)}function t(e,r){if(null==e)return{};var o,n,a=function(e,r){if(null==e)return{};var o,n,a={},t=Object.keys(e);for(n=0;n<t.length;n++)o=t[n],r.indexOf(o)>=0||(a[o]=e[o]);return a}(e,r);if(Object.getOwnPropertySymbols){var t=Object.getOwnPropertySymbols(e);for(n=0;n<t.length;n++)o=t[n],r.indexOf(o)>=0||Object.prototype.propertyIsEnumerable.call(e,o)&&(a[o]=e[o])}return a}const l={id:"error-handling",title:"Gest\xe3o de erros"},i=void 0,d={unversionedId:"Concepts/error-handling",id:"version-20-R5/Concepts/error-handling",title:"Gest\xe3o de erros",description:"O manejo de erros \xe9 o processo de antecipar e responder aos erros que possam ocorrer em sua aplica\xe7\xe3o. 4D oferece assist\xeancia completa \xe0 detec\xe7\xe3o e notifica\xe7\xe3o de erros no tempo de execu\xe7\xe3o, assim como a an\xe1lise de suas condi\xe7\xf5es.",source:"@site/i18n/pt/docusaurus-plugin-content-docs/version-20-R5/Concepts/error-handling.md",sourceDirName:"Concepts",slug:"/Concepts/error-handling",permalink:"/docs/pt/Concepts/error-handling",draft:!1,editUrl:"https://github.com/4d/docs/issues/new?title=Comment%20on%20Concepts%2Ferror-handling.md%20(20-R5)&body=Please%20enter%20your%20comment%3A",tags:[],version:"20-R5",frontMatter:{id:"error-handling",title:"Gest\xe3o de erros"},sidebar:"docs",previous:{title:"Fluxo de controlo",permalink:"/docs/pt/Concepts/control-flow"},next:{title:"Modos interpretado e compilado",permalink:"/docs/pt/Concepts/interpreted-compiled"}},s={},c=[{value:"Erro ou status",id:"erro-ou-status",level:2},{value:"Instala\xe7\xe3o de um m\xe9todo de gest\xe3o de erros",id:"instala\xe7\xe3o-de-um-m\xe9todo-de-gest\xe3o-de-erros",level:2},{value:"Alcance e componentes",id:"alcance-e-componentes",level:3},{value:"Manejo de erros dentro do m\xe9todo",id:"manejo-de-erros-dentro-do-m\xe9todo",level:3},{value:"Exemplo",id:"exemplo",level:4},{value:"Utilizar um m\xe9todo de gest\xe3o de erro vazio",id:"utilizar-um-m\xe9todo-de-gest\xe3o-de-erro-vazio",level:3},{value:"Try(expression)",id:"tryexpression",level:2},{value:"Exemplos",id:"exemplos",level:3},{value:"Try...Catch...End try",id:"trycatchend-try",level:2},{value:"Exemplo",id:"exemplo-1",level:3}],p={toc:c};function m(e){var{components:r}=e,l=t(e,["components"]);return(0,n.kt)("wrapper",a({},p,l,{components:r,mdxType:"MDXLayout"}),(0,n.kt)("p",null,"O manejo de erros \xe9 o processo de antecipar e responder aos erros que possam ocorrer em sua aplica\xe7\xe3o. 4D oferece assist\xeancia completa \xe0 detec\xe7\xe3o e notifica\xe7\xe3o de erros no tempo de execu\xe7\xe3o, assim como a an\xe1lise de suas condi\xe7\xf5es."),(0,n.kt)("p",null,"Manejo de erros responde \xe0 duas necessidades principais:"),(0,n.kt)("ul",null,(0,n.kt)("li",{parentName:"ul"},"descobrir e consertar erros potenciais e bugs no c\xf3digo durante a fase de desenvolvimento,"),(0,n.kt)("li",{parentName:"ul"},"catching and recovering from unexpected errors in deployed applications; in particular, you can replace system error dialogs (disk full, missing file, etc.) with you own interface.")),(0,n.kt)("p",null,"Basicamente, h\xe1 duas maneiras de lidar com erros em 4D. Pode:"),(0,n.kt)("ul",null,(0,n.kt)("li",{parentName:"ul"},(0,n.kt)("a",a({parentName:"li"},{href:"#installing-an-error-handling-method"}),"instalar un m\xe9todo de gesti\xf3n de errores"),", o"),(0,n.kt)("li",{parentName:"ul"},"use a ",(0,n.kt)("a",a({parentName:"li"},{href:"#tryexpression"}),(0,n.kt)("inlineCode",{parentName:"a"},"Try()")," keyword")," or a ",(0,n.kt)("a",a({parentName:"li"},{href:"#trycatchend-try"}),(0,n.kt)("inlineCode",{parentName:"a"},"Try/Catch")," structure")," before pieces of code that call a function, method, or expression that can throw an error.")),(0,n.kt)("admonition",a({},{title:"Good practice",type:"tip"}),(0,n.kt)("p",{parentName:"admonition"},">"," > \xc9 recomendado instalar um m\xe9todo de gerenciamento de erros em 4D Server, para todos os c\xf3digos rodando no servidor. Cuando 4D Server no se ejecuta ",(0,n.kt)("a",a({parentName:"p"},{href:"/docs/pt/Admin/cli"}),"headless")," (es decir, se lanza con su ",(0,n.kt)("a",a({parentName:"p"},{href:"/docs/pt/ServerWindow/overview"}),"ventana de administraci\xf3n"),"), este m\xe9todo evitar\xeda que se mostraran cajas de di\xe1logo inesperadas en la m\xe1quina servidor. En modo sin interfaz, los errores se registran en el archivo ",(0,n.kt)("a",a({parentName:"p"},{href:"/docs/pt/Debugging/debugLogFiles#4ddebuglogtxt-standard"}),"4DDebugLog")," para su posterior an\xe1lisis.")),(0,n.kt)("h2",a({},{id:"erro-ou-status"}),"Erro ou status"),(0,n.kt)("p",null,"Muchas funciones de clase 4D, como ",(0,n.kt)("inlineCode",{parentName:"p"},"entity.save()")," o ",(0,n.kt)("inlineCode",{parentName:"p"},"transporter.send()"),", devuelven un objeto ",(0,n.kt)("em",{parentName:"p"},"status"),'. Este objecto \xe9 utilizado para armazenar erros "previs\xedveis" no contexto do tempo de execu\xe7\xe3o, por exemplo, palavra-passe inv\xe1lida, entidade bloqueada, etc., que n\xe3o interrompem a execu\xe7\xe3o do programa. Esta categoria de erros pode ser tratada por c\xf3digo normal.'),(0,n.kt)("p",null,'Outros erros "imprevis\xedveis" incluem erro de grava\xe7\xe3o em disco, falha de rede, ou em geral qualquer interrup\xe7\xe3o inesperada. Esta categor\xeda de errores genera excepciones y debe gestionarse mediante un m\xe9todo de gesti\xf3n de errores o una palabra clave ',(0,n.kt)("inlineCode",{parentName:"p"},"Try()"),"."),(0,n.kt)("h2",a({},{id:"instala\xe7\xe3o-de-um-m\xe9todo-de-gest\xe3o-de-erros"}),"Instala\xe7\xe3o de um m\xe9todo de gest\xe3o de erros"),(0,n.kt)("p",null,"En 4D, todos los errores pueden ser detectados y manejados por m\xe9todos proyecto espec\xedficos, llamados ",(0,n.kt)("strong",{parentName:"p"},"m\xe9todos de gesti\xf3n de errores")," (o ",(0,n.kt)("strong",{parentName:"p"},"m\xe9todos de intercepci\xf3n de errores"),")."),(0,n.kt)("p",null,"Uma vez instalados, os manipuladores de erros s\xe3o automaticamente chamados em modo interpretado ou compilado em caso de erro na aplica\xe7\xe3o 4D ou num dos seus componentes. Um manipulador de erros diferente pode ser chamado em fun\xe7\xe3o do contexto de execu\xe7\xe3o (ver abaixo)."),(0,n.kt)("p",null,"To ",(0,n.kt)("em",{parentName:"p"},"install")," an error-handling project method, you just need to call the ",(0,n.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv19/help/command/en/page155.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"ON ERR CALL"))," command with the project method name and (optionnally) scope as parameters. Por exemplo:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("IO_ERRORS") //Instala o m\xe9todo de gest\xe3o de erros\n')),(0,n.kt)("p",null,"Para dejar de interceptar los errores en un contexto de ejecuci\xf3n y devolver la mano, llame ",(0,n.kt)("inlineCode",{parentName:"p"},"ON ERR CALL")," con una cadena vac\xeda:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("";ek local) // d\xe1 o controle para o processo local\n')),(0,n.kt)("p",null,"The  ",(0,n.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page704.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"Method called on error"))," command allows you to know the name of the method installed by ",(0,n.kt)("inlineCode",{parentName:"p"},"ON ERR CALL")," for the current process. \xc9 particularmente \xfatil no contexto dos componentes porque permite mudar temporariamente e depois restaurar o m\xe9todo de captura de erros do banco de dados local:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),' $methCurrent:=Method called on error(ek local)\n ON ERR CALL("NewMethod";ek local)\n  //If the document cannot be opened, an error is generated\n $ref:=Open document("MyDocument")\n  //Reinstallation of previous method\n ON ERR CALL($methCurrent;ek local)\n\n')),(0,n.kt)("h3",a({},{id:"alcance-e-componentes"}),"Alcance e componentes"),(0,n.kt)("p",null,"Um m\xe9todo de tratamento de erros pode ser definido para diferentes contextos de execu\xe7\xe3o:"),(0,n.kt)("ul",null,(0,n.kt)("li",{parentName:"ul"},"para el ",(0,n.kt)("strong",{parentName:"li"},"proceso actual"),"- s\xf3lo se llamar\xe1 a un gestor de errores local para los errores ocurridos en el proceso actual del proyecto actual,"),(0,n.kt)("li",{parentName:"ul"},"para ",(0,n.kt)("strong",{parentName:"li"},"toda la aplicaci\xf3n"),"- se llamar\xe1 a un gestor de errores global para todos los errores que se produzcan en el contexto de ejecuci\xf3n de la aplicaci\xf3n del proyecto actual,"),(0,n.kt)("li",{parentName:"ul"},"desde los ",(0,n.kt)("strong",{parentName:"li"},"componentes"),"- este manejador de errores se define en un proyecto local y ser\xe1 llamado para todos los errores que ocurran en los componentes cuando no hayan sido ya interceptados por un manejador de componentes.")),(0,n.kt)("p",null,"Exemplos:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("IO_Errors";ek local) //Instala um m\xe9todo local de tratamento de erros ON ERR CALL("globalHandler";ek global) //Instala um m\xe9todo global de tratamento de erros ON ERR CALL("componentHandler";ek erros de componentes) //Instala um m\xe9todo de tratamento de erros de componentes\n')),(0,n.kt)("p",null,'Pode instalar um gerenciador de erros global que servir\xe1 como "fallback" e m\xe9todos de erros locais espec\xedficos para certos processos. Um manipulador de erros global \xe9 tamb\xe9m \xfatil no servidor para evitar di\xe1logos de erro no servidor quando executado com interface.'),(0,n.kt)("p",null,"Pode definir um \xfanico m\xe9todo de captura de erros para toda a aplica\xe7\xe3o ou diferentes m\xe9todos por m\xf3dulo de aplica\xe7\xe3o. Entretanto, apenas um m\xe9todo pode ser instalado por processo."),(0,n.kt)("p",null,"Quando ocorre um erro, apenas um m\xe9todo \xe9 chamado, como descrito no diagrama seguinte:"),(0,n.kt)("p",null,(0,n.kt)("img",{alt:"error management",src:o(250161).Z,width:"1195",height:"578"})),(0,n.kt)("h3",a({},{id:"manejo-de-erros-dentro-do-m\xe9todo"}),"Manejo de erros dentro do m\xe9todo"),(0,n.kt)("p",null,"Within the custom error method, you have access to several pieces of information that will help you identifying the error:"),(0,n.kt)("ul",null,(0,n.kt)("li",{parentName:"ul"},(0,n.kt)("p",{parentName:"li"},"vari\xe1veis sistema (","*","):"),(0,n.kt)("ul",{parentName:"li"},(0,n.kt)("li",{parentName:"ul"},(0,n.kt)("inlineCode",{parentName:"li"},"Error")," (entero largo): c\xf3digo de error"),(0,n.kt)("li",{parentName:"ul"},(0,n.kt)("inlineCode",{parentName:"li"},"Error method"),"(texto): nombre del m\xe9todo que ha provocado el error"),(0,n.kt)("li",{parentName:"ul"},(0,n.kt)("inlineCode",{parentName:"li"},"Error line")," (entero largo): n\xfamero de l\xednea del m\xe9todo que ha provocado el error"),(0,n.kt)("li",{parentName:"ul"},(0,n.kt)("inlineCode",{parentName:"li"},"Error formula")," (texto): f\xf3rmula del c\xf3digo 4D (texto bruto) que est\xe1 en el origen del error.")))),(0,n.kt)("admonition",a({},{type:"info"}),(0,n.kt)("p",{parentName:"admonition"},"4D mantiene autom\xe1ticamente una serie de variables denominadas ",(0,n.kt)("strong",{parentName:"p"},"variables sistema"),", que responden a diferentes necesidades. Ver o manual da linguagem 4D","*",".")),(0,n.kt)("ul",null,(0,n.kt)("li",{parentName:"ul"},"the ",(0,n.kt)("a",a({parentName:"li"},{href:"https://doc.4d.com/4dv19/help/command/en/page1799.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"Last errors"))," command that returns a collection of the current stack of errors that occurred in the 4D application. You can also use the ",(0,n.kt)("a",a({parentName:"li"},{href:"https://doc.4d.com/4dv19/help/command/en/page1015.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"GET LAST ERROR STACK"))," command that returns the same information as arrays."),(0,n.kt)("li",{parentName:"ul"},"el comando ",(0,n.kt)("inlineCode",{parentName:"li"},"Get call chain")," que devuelve una colecci\xf3n de objetos que describen cada paso de la cadena de llamadas a m\xe9todos dentro del proceso actual.")),(0,n.kt)("h4",a({},{id:"exemplo"}),"Exemplo"),(0,n.kt)("p",null,"Aqui est\xe1 um sistema de gest\xe3o de erros simples:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'//instalar o m\xe9todo de gest\xe3o de erros\n ON ERR CALL("errorMethod")\n //... executar o c\xf3digo\n ON ERR CALL("") //retorna o controle para 4D\n')),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'// errorMethod project method\n If(Error#1006) //this is not a user interruption\n    ALERT("The error "+String(Error)+" occurred". The code in question is: \\""+Error formula+"\\"")\n End if\n')),(0,n.kt)("h3",a({},{id:"utilizar-um-m\xe9todo-de-gest\xe3o-de-erro-vazio"}),"Utilizar um m\xe9todo de gest\xe3o de erro vazio"),(0,n.kt)("p",null,"Se quiser que a caixa de di\xe1logo fique escondida, pode instalar um m\xe9todo de gest\xe3o de erros vazio. La variable sistema ",(0,n.kt)("inlineCode",{parentName:"p"},"Error")," puede ser probada en cualquier m\xe9todo, es decir, fuera del m\xe9todo de gesti\xf3n de errores:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("emptyMethod") //emptyMethod exists but is empty\n$doc:=Open document( "myFile.txt")\nIf (Error=-43)\n    ALERT("File not found.")\nEnd if\nON ERR CALL("")\n')),(0,n.kt)("h2",a({},{id:"tryexpression"}),"Try(expression)"),(0,n.kt)("p",null,"La sentencia ",(0,n.kt)("inlineCode",{parentName:"p"},"Try(expression)")," permite probar una expresi\xf3n de una sola l\xednea en su contexto de ejecuci\xf3n real (incluyendo, en particular, los valores de las variables locales) e interceptar los errores que arroje para que no se muestre el di\xe1logo de erro El uso de ",(0,n.kt)("inlineCode",{parentName:"p"},"Try(expression)")," ofrece una manera f\xe1cil de manejar casos de error simples con un n\xfamero muy bajo de l\xedneas de c\xf3digo, y sin requerir un m\xe9todo de gesti\xf3n de errores."),(0,n.kt)("admonition",a({},{type:"note"}),(0,n.kt)("p",{parentName:"admonition"},"If you want to try a more complex code than a single-line expression, you might consider using a ",(0,n.kt)("a",a({parentName:"p"},{href:"#trycatchend-try"}),(0,n.kt)("inlineCode",{parentName:"a"},"Try/Catch")," structure"),".")),(0,n.kt)("p",null,"La sintaxis formal de la declaraci\xf3n ",(0,n.kt)("inlineCode",{parentName:"p"},"Try(expresi\xf3n)")," es:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),"\nTry (expression) : any | Undefined\n\n")),(0,n.kt)("p",null,(0,n.kt)("em",{parentName:"p"},"expresion")," puede ser toda expresi\xf3n v\xe1lida."),(0,n.kt)("p",null,"Si se produce un error durante su ejecuci\xf3n, se intercepta y no se muestra ning\xfan di\xe1logo de error, si un ",(0,n.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"m\xe9todo de gesti\xf3n de errores")," fue instalado o no antes de la llamada a ",(0,n.kt)("inlineCode",{parentName:"p"},"Try()"),". Si ",(0,n.kt)("em",{parentName:"p"},"expression")," devuelve un valor, ",(0,n.kt)("inlineCode",{parentName:"p"},"Try()")," devuelve el \xfaltimo valor evaluado, en caso contrario devuelve ",(0,n.kt)("inlineCode",{parentName:"p"},"Define"),"."),(0,n.kt)("p",null,"You can handle the error(s) using the ",(0,n.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page1799.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"Last errors"))," command. Si ",(0,n.kt)("em",{parentName:"p"},"expression")," arroja un error dentro de una pila de llamadas ",(0,n.kt)("inlineCode",{parentName:"p"},"Try()"),", el flujo de ejecuci\xf3n se detiene y devuelve a la \xfaltima ejecuci\xf3n ",(0,n.kt)("inlineCode",{parentName:"p"},"Try()")," (la primera encontrada de nuevo en la pila de llamadas)."),(0,n.kt)("admonition",a({},{type:"note"}),(0,n.kt)("p",{parentName:"admonition"},"Si un ",(0,n.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"m\xe9todo de gesti\xf3n de errores")," es instalado por ",(0,n.kt)("em",{parentName:"p"},"expression"),", es llamado en caso de error.")),(0,n.kt)("h3",a({},{id:"exemplos"}),"Exemplos"),(0,n.kt)("ol",null,(0,n.kt)("li",{parentName:"ol"},"You want to display the contents of a file if the file can be open without error, and if its contents can be read. Voc\xea pode escrever:")),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'var $text : Text\nvar $file : 4D.File := File("/RESOURCES/myFile.txt")\nvar $fileHandle : 4D.FileHandle := Try($file.open())\nIf ($fileHandle # Null)\n  $text:=Try($fileHandle.readText()) || "Erro ao ler o arquivo"\nEnd if\n')),(0,n.kt)("ol",a({},{start:2}),(0,n.kt)("li",{parentName:"ol"},"Voc\xea deseja tratar o erro de divis\xe3o por zero. Nesse caso, voc\xea deseja retornar 0 e lan\xe7ar um erro:")),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'function divide( $p1: real; $p2: real)-> $result: real\n  if ($p2=0)\n     $result:=0 //only for clarity (0 is the default for reals)\n     throw(-12345; "Division by zero")\n  else\n    $result:=$p1/$p2\n  end if\n\nfunction test()\n  $result:=Try(divide($p1;$p2))\n  If (Last errors # null)\n    ALERT("Error")\n  End if\n\n')),(0,n.kt)("ol",a({},{start:3}),(0,n.kt)("li",{parentName:"ol"},"Desea gestionar tanto los errores ",(0,n.kt)("a",a({parentName:"li"},{href:"#error-or-status"}),"previsibles como los no previsibles>"),":")),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'var $e:=ds.Employee.new()\n$e.name:="Smith"\n$status:=Try($e.save()) //catch predictable and non-predictable errors\nIf ($status.success)\n   ALERT( "Success")\nElse\n   ALERT( "Error: "+JSON Stringify($status.errors))\nEnd if\n\n')),(0,n.kt)("h2",a({},{id:"trycatchend-try"}),"Try...Catch...End try"),(0,n.kt)("p",null,"La estructura ",(0,n.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," permite probar el c\xf3digo de un bloque en su contexto de ejecuci\xf3n real (incluyendo, en particular, los valores de las variables locales) e interceptar los errores que lanza para que no se muestre el di\xe1logo de error d"),(0,n.kt)("p",null,"A diferencia de la palabra clave ",(0,n.kt)("inlineCode",{parentName:"p"},"Try(expression)")," que eval\xfaa una expresi\xf3n de una sola l\xednea, la estructura ",(0,n.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," permite evaluar cualquier bloque de c\xf3digo, desde el m\xe1s simple al m\xe1s complejo, sin necesidad de un m\xe9todo de gesti\xf3n de  Adem\xe1s, el bloque ",(0,n.kt)("inlineCode",{parentName:"p"},"Catch")," puede utilizarse para gestionar el error de forma personalizada."),(0,n.kt)("p",null,"La sintaxis formal de la estructura ",(0,n.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," es:"),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),"\nTry \n    statement(s) // Code to evaluate\nCatch\n    statement(s) // Code to execute in case of error\nEnd try\n\n")),(0,n.kt)("p",null,"El c\xf3digo entre las palabras clave ",(0,n.kt)("inlineCode",{parentName:"p"},"Try")," y ",(0,n.kt)("inlineCode",{parentName:"p"},"Catch")," se ejecuta en primer lugar, luego el flujo depende del error o errores encontrados durante esta ejecuci\xf3n."),(0,n.kt)("ul",null,(0,n.kt)("li",{parentName:"ul"},"Si no se lanza ning\xfan error, la ejecuci\xf3n del c\xf3digo contin\xfaa despu\xe9s de la palabra clave correspondiente ",(0,n.kt)("inlineCode",{parentName:"li"},"End try"),". El c\xf3digo situado entre las palabras clave ",(0,n.kt)("inlineCode",{parentName:"li"},"Catch")," y ",(0,n.kt)("inlineCode",{parentName:"li"},"End try")," no se ejecuta."),(0,n.kt)("li",{parentName:"ul"},"Si la ejecuci\xf3n del bloque de c\xf3digo arroja un error no diferido ","*","*",", el flujo de ejecuci\xf3n se detiene y ejecuta el bloque de c\xf3digo correspondiente ",(0,n.kt)("inlineCode",{parentName:"li"},"Catch"),"."),(0,n.kt)("li",{parentName:"ul"},"Si la ejecuci\xf3n del bloque de c\xf3digo arroja un ",(0,n.kt)("em",{parentName:"li"},"error diferido"),", el flujo de ejecuci\xf3n contin\xfaa hasta el final del bloque ",(0,n.kt)("inlineCode",{parentName:"li"},"Try")," y luego ejecuta el bloque de c\xf3digo correspondiente ",(0,n.kt)("inlineCode",{parentName:"li"},"Catch"),".")),(0,n.kt)("admonition",a({},{type:"note"}),(0,n.kt)("p",{parentName:"admonition"},"Si se lanza un error ",(0,n.kt)("em",{parentName:"p"},"diferido")," fuera del bloque ",(0,n.kt)("inlineCode",{parentName:"p"},"Try"),", la ejecuci\xf3n del c\xf3digo contin\xfaa hasta el final del m\xe9todo o funci\xf3n.")),(0,n.kt)("admonition",a({},{type:"info"}),(0,n.kt)("p",{parentName:"admonition"},"For more information on ",(0,n.kt)("em",{parentName:"p"},"deferred")," and ",(0,n.kt)("em",{parentName:"p"},"non-deferred")," errors, please refer to the ",(0,n.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20R/help/command/en/page1805.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"throw"))," command description.")),(0,n.kt)("p",null,"En el bloque de c\xf3digo ",(0,n.kt)("inlineCode",{parentName:"p"},"Catch"),", puede gestionar los errores utilizando los comandos est\xe1ndar de gesti\xf3n de errores. The ",(0,n.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page1799.html"}),(0,n.kt)("inlineCode",{parentName:"a"},"Last errors"))," function contains the last errors collection. En este bloque de c\xf3digo puede declarar ",(0,n.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"un m\xe9todo de gesti\xf3n de errores"),", en cuyo caso se llama en caso de error (de lo contrario se muestra el di\xe1logo de error de 4D)."),(0,n.kt)("admonition",a({},{type:"note"}),(0,n.kt)("p",{parentName:"admonition"},"Si se instala un ",(0,n.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"m\xe9todo de gesti\xf3n de errores")," en el c\xf3digo colocado entre las palabras clave ",(0,n.kt)("inlineCode",{parentName:"p"},"Try")," y ",(0,n.kt)("inlineCode",{parentName:"p"},"Catch"),", se llama en caso de error.")),(0,n.kt)("h3",a({},{id:"exemplo-1"}),"Exemplo"),(0,n.kt)("p",null,"La combinaci\xf3n de transacciones y estructuras ",(0,n.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," permite escribir c\xf3digo seguro para funciones cr\xedticas."),(0,n.kt)("pre",null,(0,n.kt)("code",a({parentName:"pre"},{className:"language-4d"}),"Function createInvoice($customer : cs.customerEntity; $items : Collection; $invoiceRef : Text) : cs.invoiceEntity\n    var $newInvoice : cs.invoiceEntity\n    var $newInvoiceLine : cs.invoiceLineEntity\n    var $item : Object\n    ds.startTransaction()\n    Try\n        $newInvoice:=This.new()\n        $newInvoice.customer:=$customer\n        $newInvoice.invoiceRef:=$invoiceRef\n        For each ($item; $items)\n            $newInvoiceLine:=ds.invoiceLine.new()\n            $newInvoiceLine.item:=$item.item\n            $newInvoiceLine.amount:=$item.amount\n            $newInvoiceLine.invoice:=$newInvoice\n            //call other specific functions to validate invoiceline\n            $newInvoiceLine.save()\n        End for each \n        $newInvoice.save()\n        ds.validateTransaction()\n    Catch\n        ds.cancelTransaction()\n        ds.logErrors(Last errors)\n        $newInvoice:=Null\n    End try\n    return $newInvoice\n\n")))}m.isMDXComponent=!0},250161:(e,r,o)=>{o.d(r,{Z:()=>n});const n=o.p+"assets/images/error-schema-973d40c4618e79de87f884622e4e930f.png"}}]);